gui/quantum.lua

```diff
@@ -11,262 +11,218 @@ local quickfort = reqscript('quickfort')
 end
 
-function QuantumUI:on_back()
-    if self.feeder then
-        self.feeder = nil
-        self:updateLayout()
-    else
-        self:dismiss()
-    end
+function QuantumWin:on_back()
+    self:dismiss()
 end
 
-function QuantumUI:on_name_char(char, text)
-    return #text < 12
+function QuantumUI:on_back()
+    self:dismiss()
 end
 
-local function is_in_extents(bld, x, y)
-    local extents = bld.room.extents
-    if not extents then return true end -- building is solid
-    local yoff = (y - bld.y1) * (bld.x2 - bld.x1 + 1)
-    local xoff = x - bld.x1
-    return extents[yoff+xoff] == 1
+function QuantumWin:on_name_char(char, text)
+    return #text < 12
 end
 
-function QuantumUI:select_stockpile(pos)
-    local flags, occupancy = dfhack.maps.getTileFlags(pos)
-    if not flags or occupancy.building == 0 then return end
-    local bld = dfhack.buildings.findAtTile(pos)
-    if not bld or bld:getType() ~= df.building_type.Stockpile then return end
-
-    local tiles = {}
+--UI Data Tables
+qspOutputHor = {
+    ["North"]=0,
+    ["East"]=1,
+    ["South"]=0,
+    ["West"]=-1
+}
 
-    for x=bld.x1,bld.x2 do
-        for y=bld.y1,bld.y2 do
-            if is_in_extents(bld, x, y) then
-                ensure_key(ensure_key(tiles, bld.z), y)[x] = true
-            end
-        end
-    end
+qspOutputVert = {
+    ["North"]=-1,
+    ["East"]=0,
+    ["South"]=1,
+    ["West"]=0
+}
 
-    self.feeder = bld
-    self.feeder_tiles = tiles
+qspTrackstopDir = {
+    ["North"]='trackstopN',
+    ["East"]='trackstopE',
+    ["South"]='trackstopS',
+    ["West"]='trackstopW'
+}
 
-    self:updateLayout()
-end
+qspOutputType = {
+    [true]='ry',
+    [false]='c'
+}
 
-function QuantumUI:render_feeder_overlay()
-    if not gui.blink_visible(1000) then return end
 
-    local zlevel = self.feeder_tiles[df.global.window_z]
-    if not zlevel then return end
 
-    local function get_feeder_overlay_char(pos)
-        return safe_index(zlevel, pos.y, pos.x) and 'X'
+--Actually making the QSP
+function QuantumWin:commit()
+    name = self.subviews.name.text
+    dump_dir = self.subviews.dir:getOptionLabel()
+    allow_refuse = self.subviews.refuse:getOptionValue()
+    input_id = dfhack.gui.getSelectedStockpile().id
+    output_hor = qspOutputHor[dump_dir]
+    output_vert = qspOutputVert[dump_dir]
+    output_type = qspOutputType[allow_refuse]
+    trackstop_dir = qspTrackstopDir[dump_dir]
+    if
+        name == ''
+    then
+        name = 'Unnamed QSP'
     end
-
-    self:renderMapOverlay(get_feeder_overlay_char, self.feeder)
-end
-
-function QuantumUI:get_qsp_pos(cursor)
-    local offsets = self.subviews.dir:getOptionValue()
-    return {
-        x = cursor.x + (offsets.x or 0),
-        y = cursor.y + (offsets.y or 0),
-        z = cursor.z
+    placed_output = output_type..'{name="'..name..'" quantum=true}:+all'
+    placed_trackstop = trackstop_dir..'{name="'..name..'" take_from='..input_id..'}'
+
+    quickfort.apply_blueprint{
```

Comment:
this needs the error checking that was in the original code -- if the player has selected an invalid tile for the QSP, we should not blindly move forward

Reply:
This was somewhat of a deliberate design choice. There are a few uses for a pseudo QSP where the output tile isn't valid to put a stockpile on, for example, dumping goblin corpses down a mass grave out of sight and out of mind, dumping ores down a huge shaft down to the magma sea and magma forges, and dumping iron minecarts onto a roller system to fill them with magma. I've even used quantum to make storm drains and water reactors via a dummy 1x1 input stockpile with no settings, since they both use the same mechanism of a dumping Minecart.

As for the error checking, wouldn't it be redundant? From what I could tell from the old version, the error checking went place trackstop (dry run), if dry placed entities = 0 then display error message, else place trackstop. What's stopping us from place trackstop, if placed entities = 0 then display error message? I could see it being necessary for the trackstop itself, rather than the output stockpile because regardless of the above niche uses or an actual QSP, neither will work without the trackstop, unless it causes problems in the backend I'm not aware of.